### DISCUSSION TITLE
Onboarding

### DISCUSSION THREAD

**Discussion by jmoldyvan** [timmyichen/dev-directory#23] [Technical Specification Review]
**Summary**:

  This spec goes into the idea of the best way to onboard users.
  Ideally, we want to provide a friendly welcome for new users and help them to pair with other programmers/users for collaboration. It also includes signing in process and features exploration.

**Scope**:

  
  >* Initial welcome page with login using our authentication
  >* Integrating developer profile/user data to create onboarding questions using simple UI forms
  >* Check to see if the user has previously been onboarded or not with a flags table
  
**Out of Scope**:

  >* Determine the specifics of Onboarding diction
  >* Improve the design of UI, make it visually friendly
  >* Integrate other potential profile data like role, experience, etc
  >* Update the onBoarding page and form to include multiple pages containing individual forms
  >* Adding a progress bar along with the multi-page scheme

**Data model**: N/A

**User Stories**:

  >* Users will be greeted
  >* Users will be able to create all or some of their bio 
  >* All questions and onBoarding should be skippable

**Features**:

      -Entrance Flow-
      
      1. Greeting/user signs up using Discord OAuth
      3. If the user has been through onboarding, they go directly to the dev-directory list of profiles page
      4. If user has not been through onboarding they are redirected to the onboarding page
      5. Onboarding page which includes a form (filling out some or all of the profile which is currently the `bio`, and social links)
      6. If the onboarding page is skipped at any time or if the user submits their forms they will be redirected users to the dev-directory list of profiles page

**New pages/views**:

  >*  Greeting page including the Discord Auth
  >* A singular page containing a form with the following questions:
  >* "tell us a bit about yourself" `bio`
  >* "what are your social links?" `twitter_username`, `github_username`, `linkedin_url`, `website`
  >* Include forms for any other data we want to create at the initial point of onboarding

**Forms to Include**:

  >* Discord auth i.e. log in (button)
  >* Bio (string, max 1000 chars)
  >* Social Links: 
Twitter, Discord, Github, website  (string, max 200 chars)

**APIs**:

 This is the route we will use to update user fields.
 
    1.  PATCH /api/users/details - updates a subset of user profile columns, applied to the currently-logged-in-user.

**Security**: N/A

**Testing**:

  >* Making sure the forms are updating the server/db through endpoints.
  >* Making sure users are logged in or signed up when they complete the form. 
    

**Open Questions**:

  >* Is having a multipage onboarding process too much for an MVP? Would a single page, with multiple forms make more sense?
  >* There was an idea of using a modal format during a tech spec review meeting, do we want to have the onboarding forms be a modal or a separate page? #61
  >* We need a way to keep track of users who have already seen the onbaording previously. How should we approach this? #63 

**Deployment**:

  
  >* Test features

**Comment by timmyichen** [comment]
> Users will be able to login

Users are already able to login so I don't think this is necessarily a user story for this project.

For the open questions:

> Which page do we want users to see first after onboarding?

I can see us prompting them to fill out some of their profile. There's research that shows that giving them multiple steps makes it more likely that they'll fill it out (vs asking them to fill out a longer form).  I think we can split it up into "tell us a bit about yourself" page (bio) and "what are your social links?" (social links).

> Blocking users from using app until certain forms are filled in?

All onboarding should be skip-able, though maybe we should have some kind of progress bar when they log in so they're encouraged to finish?

> Log in and send an email welcome?

No need for email, I think, partially because we don't want to pay for an emailing service :P

> Two separate forms for mentor or mentee?

Given we don't have mentor/mentee filters yet, I think we can do without them - 

> Do we want multiple forms/pages or one large form one a single page?

See above - multiple pages with smaller, manageable pieces results in higher engagement vs not.

> How to use app (video or explanation)

Ideally this app is self explanatory, but I also think that teaching users how to use the app is outside the scope of this project.

For the other forms, I think we can exclude them, as we just want to onboard them with the user fields we have right now - though we can certainly do it later as we add more user fields!

**Comment by Caleb-Cohen** [comment]
> Which page do we want users to see first after onboarding?

I'd redirect the user directly to the directory itself. Let them get right into using the product.

> Blocking users from using app until certain forms are filled in?

Agreed with @timmyichen if the user doesn't want to answer something, just let them skip. We should have all the necessary information on login from Discord.

> Log in and send an email welcome?

Works for me!

> Two separate forms for mentor or mentee?

What would this accomplish? My thought would be to have a question like: Are you interested in mentoring on being a mentee?

> Do we want multiple forms/pages or one large form one a single page?

forms should be broken up into categories. Makes it more manageable for the user.

> How to use app (video or explanation)

I'd like to make it as intuitive as possible. Video can come later if we have the user base and complexity to be valuable.

**Reply by timmyichen** [reply]
> I'd redirect the user directly to the directory itself. Let them get right into using the product.

They can skip onboarding if they wanted to, but not prompting them at the beginning makes it less likely for them to fill it out later - you'll end up with most of your users having empty profiles

**Reply by Caleb-Cohen** [reply]
> > I'd redirect the user directly to the directory itself. Let them get right into using the product.
> 
> They can skip onboarding if they wanted to, but not prompting them at the beginning makes it less likely for them to fill it out later - you'll end up with most of your users having empty profiles

Yeah I agree with that!

**Comment by jmoldyvan** [comment]
>*Log in and send an email welcome?

if this costs money than we can trash it!

>*Which page do we want users to see first after onboarding?

so what directory page? I guess what is our main page called?

>*Blocking users from using app until certain forms are filled in?

I also agree the skippable aspect of the onbaording seems good

>*Two separate forms for mentor or mentee?

I agree I think a prompt early would be best

>*Do we want multiple forms/pages or one large form one a single page?

yes multiple pages does seem more manageable by users

**Reply by jmoldyvan** [reply]
I also updated some of the tech spec to reflect this. but also asked another open question

**Comment by timmyichen** [comment]
Thanks for making those edits!  The additional open questions:

> Which page do we want users to see first after onboarding?

I think it makes sense to do user login -> if not onboarded, user gets greeted -> goes to the relevant onboarding step -> goes to browse profiles page

> Should we include a progress bar for completed forms?

I think this would be nice to have!

**Comment by jmoldyvan** [comment]
I updated the spec to include open questions regarding some of the tickets I created. Let me know what you think, or if we need to discuss this further during a meeting. Happy to make adjustments either way/whenever we figure this out.

**Comment by timmyichen** [comment]
Thanks for making the additions!

My opinions are:

I like the idea of a progress bar!

Pages over modals. People dismiss modals instinctively and we _want_ people to finish onboarding.

Multiple pages over a single page. Makes it more likely for people to fill out, and more likely that people will partially fill out before stopping. It will also be easier to chain a bunch of pages together over a bunch of modals, and is better ux too.  What are the benefits of using a modal?

Re: how to tell if something has finished onboarding, we can use a few flags table as described in one of the closed tickets (I'm on mobile, forgot which one). Can include flags for each of the steps, or a single flag for the entire thing. Benefits of multiple are we can add onboarding steps and still put people through the new flow.
